pkg/controller/datavolume-controller.go

```diff
@@ -595,24 +553,194 @@ func (r *DatavolumeReconciler) reconcileSmartClonePvc(log logr.Logger, datavolum
 				return reconcile.Result{}, err
 			}
 		} else if tmpPVC.Annotations[AnnCloneOf] == "true" {
-			done, err := r.expand(log, datavolume, tmpPVC, pvcSpec)
-			if err != nil {
-				return reconcile.Result{}, err
-			}
-			if !done {
-				return reconcile.Result{},
-					r.updateSmartCloneStatusPhase(cdiv1.ExpansionInProgress, datavolume, nil)
-			}
+			return r.expandTmpPvcAfterClone(log, datavolume, pvcSpec, tmpPVC, CsiClone)
+		}
+	}
 
-			// trigger transfer and next reconcile should have pvcExists == true
-			tmpPVC.Annotations[annReadyForTransfer] = "true"
-			tmpPVC.Annotations[AnnPopulatedFor] = datavolume.Name
-			if err := r.client.Update(context.TODO(), tmpPVC); err != nil {
-				return reconcile.Result{}, err
+	if datavolume.Status.Phase == cdiv1.NamespaceTransferInProgress {
+		return reconcile.Result{}, nil
+	}
+
+	sourcePvcNs := datavolume.Spec.Source.PVC.Namespace
+	if sourcePvcNs == "" {
+		sourcePvcNs = datavolume.Namespace
+	}
+	r.log.V(3).Info("CSI-Clone is available")
+
+	// Get source pvc
+	sourcePvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: sourcePvcNs, Name: datavolume.Spec.Source.PVC.Name}, sourcePvc); err != nil {
+		if k8serrors.IsNotFound(err) {
+			log.V(3).Info("Source PVC no longer exists")
+			return reconcile.Result{}, err // r.updateCSICloneDataVolumeStatus(cdiv1.CloneSourcePVCLost, datavolume)
+		}
+		return reconcile.Result{}, err
+	}
+
+	inUse, err := r.sourceInUse(datavolume, CSICloneSourceInUse)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	populated, err := r.isSourcePVCPopulated(datavolume)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	if inUse || !populated {
+		return reconcile.Result{Requeue: true},
+			r.updateCloneStatusPhase(cdiv1.CloneScheduled, datavolume, nil, CsiClone)
+	}
+
+	// here use Temp PVC!!!
+	cloneTargetPvc := newVolumeClonePVC(datavolume, sourcePvc, pvcSpec, pvcName)
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Namespace: cloneTargetPvc.Namespace, Name: cloneTargetPvc.Name}, pvc); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		if err := r.client.Create(context.TODO(), cloneTargetPvc); err != nil && !k8serrors.IsAlreadyExists(err) {
+			return reconcile.Result{}, err
+		}
+	} else {
+		// PVC already exists, check for name clash
+		pvcControllerRef := metav1.GetControllerOf(cloneTargetPvc)
+		pvcClashControllerRef := metav1.GetControllerOf(pvc)
+
+		if pvc.Name == cloneTargetPvc.Name &&
+			pvc.Namespace == cloneTargetPvc.Namespace &&
+			!reflect.DeepEqual(pvcControllerRef, pvcClashControllerRef) {
+			return reconcile.Result{}, errors.Errorf("Target Pvc Name in use") //r.updateCSICloneDataVolumeStatus(cdiv1.ClonePVCNameInUse, datavolume)
+		}
+
+		if pvc.Status.Phase == corev1.ClaimBound {
+			if v, ok := pvc.Annotations[AnnCloneOf]; !ok || v != "true" {
```

Comment:
I see this code several times, maybe we can make it a function, like 'setCloneOfOnPvc'?

Reply:
Done